tests/Bus/TestBusTest.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace Zenstruck\Messenger\Test\Tests\Bus;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\Stamp\StampInterface;
+use Zenstruck\Messenger\Test\Bus\TestBus;
+
+class TestBusTest extends TestCase
+{
+    /**
+     * @test
+     */
+    public function collect_messages_by_default(): void
+    {
+        $bus = new TestBus('bus', $mock = $this->createMock(MessageBusInterface::class));
+        $message = new \stdClass();
+        $mock->expects(self::once())->method('dispatch')
+            ->with(self::isInstanceOf(\stdClass::class), [])
+            ->willReturn(new Envelope($message));
+
+        $bus->dispatch($message);
+
+        $bus->dispatched()->assertCount(1);
+        $bus->dispatched()->assertContains(\stdClass::class, 1);
+    }
+
+    /**
+     * @test
+     */
+    public function collect_messages_when_enabled(): void
+    {
+        $bus = new TestBus('bus', new TestableBus());
+
+        $bus->dispatch(new \stdClass(), [$this->createMock(StampInterface::class)]);
+        TestBus::disableMessagesCollection();
+        $bus->dispatch(new \stdClass(), [$this->createMock(StampInterface::class)]);
+
+        $bus->dispatched()->assertCount(1);
+        self::assertCount(2, TestableBus::$envelopes);
+    }
+
+    /**
+     * @test
+     */
+    public function reset_messages(): void
+    {
+        $bus = new TestBus('bus', $mock = $this->createMock(MessageBusInterface::class));
+        $mock->method('dispatch')->willReturn(new Envelope(new \stdClass()));
+
+        $bus->dispatch(new \stdClass());
+
+        TestBus::resetAll();
+        $bus->dispatched()->assertEmpty();
+    }
+}
+
+final class TestableBus implements MessageBusInterface
+{
+    static array $envelopes = [];
```

Comment:
no need for this to be static (if used in multiple tests, it would have a bug) 
```suggestion
    public array $envelopes = [];
```